### PR TITLE
[SPARK-47777][PYTHON][SS][TESTS] Add spark connect test for python streaming data source

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Make python streaming data source pyspark test also runs on spark connect. Refactor the test because foreachbatch runs on spark connect server and do not update local variable.

### Why are the changes needed?
Test python streaming data source on spark connect.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Test change.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: python/pyspark/sql/tests/connect/test_parity_python_streaming_datasource.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from pyspark.sql.tests.test_python_streaming_datasource import BasePythonStreamingDataSourceTestsMixin
+from pyspark.testing.connectutils import ReusedConnectTestCase
+
+
+class PythonStreamingDataSourceParityTests(BasePythonStreamingDataSourceTestsMixin, ReusedConnectTestCase):
+    ...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_parity_python_streaming_datasource.py:22]: Let's use `pass`

File: python/pyspark/sql/tests/connect/test_parity_python_streaming_datasource.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_parity_python_streaming_datasource.py:1]: should list this file into https://github.com/apache/spark/blob/master/dev/sparktestsupport/modules.py

Comment (chaoqin-li1123) [reply python/pyspark/sql/tests/connect/test_parity_python_streaming_datasource.py:1]: Added.

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
LGTM once CI pass, thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by chaoqin-li1123** [comment]
cc @allisonwang-db @HeartSaVioR

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @chaoqin-li1123 and all.
Merged to master for Apache Spark 4.0.0.

**Comment by HyukjinKwon** [comment]
@chaoqin-li1123 Seems like this test does not work with pure Python library. Can you see if the tests pass after removing `python/lib/py4j*.zip`?

Let me revert this for now because we're cutting the preview very soon.

**Comment by chaoqin-li1123** [comment]
@HyukjinKwon both test_python_datasource, test_python_streaming_datasource will fail with the same error if py4j*.zip is removed. 

> Traceback (most recent call last):
>   File "<frozen runpy>", line 189, in _run_module_as_main
>   File "<frozen runpy>", line 112, in _get_module_details
>   File "/Users/chaoqin.li/spark/python/pyspark/__init__.py", line 58, in <module>
>     from pyspark.core.status import StatusTracker, SparkJobInfo, SparkStageInfo
>   File "/Users/chaoqin.li/spark/python/pyspark/core/status.py", line 22, in <module>
>     from py4j.java_collections import JavaArray
> ModuleNotFoundError: No module named 'py4j'

**Comment by HyukjinKwon** [comment]
py4j shouldn't be referred for connect test. can we move them, and import when it's actually used?

**Comment by HyukjinKwon** [comment]
follow https://github.com/apache/spark/blob/master/.github/workflows/build_python_connect.yml#L80-L113 to reproduce the failure

**Comment by chaoqin-li1123** [comment]
This seems to be broken in the main function of pyspark init(), what is the expected action item we should take? @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Should follow https://github.com/apache/spark/blob/master/.github/workflows/build_python_connect.yml#L69-L94 steps, and try to reproduce why it fails. Spark Connect doesn't need to import `pyspark` but only `pyspark.sql`.

**Comment by HyukjinKwon** [comment]
@chaoqin-li1123 gentle ping on this.

**Comment by chaoqin-li1123** [comment]
This import statement is supposed to be skipped for spark connect test https://github.com/apache/spark/blob/master/python/pyspark/__init__.py#L55
Is the is_remote_only() function working correctly?
@HyukjinKwon

**Comment by HyukjinKwon** [comment]
```
    from pyspark.core.rdd import RDD, RDDBarrier
    from pyspark.core.files import SparkFiles
    from pyspark.core.status import StatusTracker, SparkJobInfo, SparkStageInfo
    from pyspark.core.broadcast import Broadcast
    from pyspark.core import rdd, files, status, broadcast
```

are not supported in pure Python library for Spark Connect so we shouldn't use them for Spark Connect related modules.

**Comment by chaoqin-li1123** [comment]
Yes, but these imports are wrapped in if not is_remote_only(), so spark connect test should already skip these import statement. This is weird, I will take another look.

**Pull request closed without merging**